molviewspec/molviewspec/nodes.py

```diff
@@ -591,29 +610,94 @@ class CanvasParams(BaseModel):
-    color: ColorT = Field(description="Color of the sphere")
-    label: Optional[str] = Field(description="Optional text label")
-    tooltip: Optional[str] = Field(description="Optional tooltip label, shown upon hover")
 
+class PrimitivesParams(BaseModel):
+    default_color: Optional[ColorT] = Field(description="Default color for primitives in this group")
+    default_label_color: Optional[ColorT] = Field(description="Default label color for primitives in this group")
+    default_tooltip: Optional[str] = Field(description="Default tooltip for primitives in this group")
+    transparency: Optional[float] = Field(description="Transparency of primitives in this group")
 
-class LineParams(BaseModel):
+
+class MeshParams(BaseModel):
     """
-    Experimental: Defines a line primitive.
+    Low-level, fully customizable mesh representation of a shape.
     """
 
-    position1: tuple[float, float, float] = Field(description="Start position of the primitive")
-    position2: tuple[float, float, float] = Field(description="End position of the primitive")
-    radius: float = Field(description="Radius of the line")
-    color: ColorT = Field(description="Color of the line")
-    label: Optional[str] = Field(description="Optional text label")
-    tooltip: Optional[str] = Field(description="Optional tooltip label, shown upon hover")
+    kind: Literal["mesh"] = "mesh"
+    vertices: list[float] = Field(description="3N length array of floats with vertex position (x1, y1, z1, ...)")
+    indices: list[int] = Field(
+        description="3N length array of indices into vertices that form triangles (t1_1, t1_2, t1_3, ...)"
+    )
+    triangle_groups: Optional[list[int]] = Field(description="Assign a number to each triangle to group them.")
+    group_colors: Optional[Mapping[int, ColorT]] = Field(
+        description="Assign a color to each group. If not assigned, default primitives group color is used. Takes precedence over triangle_colors."
+    )
+    group_tooltips: Optional[Mapping[int, str]] = Field(description="Assign an optional tooltip to each group.")
+    triangle_colors: Optional[list[ColorT]] = Field(description="Assign a color to each triangle.")
+    tooltip: Optional[str] = Field(
+        description="Tooltip shown when hovering over the mesh. Assigned group_tooltips take precedence."
+    )
+
+
+class CircleParams(BaseModel):
+    kind: Literal["circle"] = "circle"
+    center: PositionT = Field(description="Center of circle.")
+    radius: float = Field(description="Radius of circle.", gt=0.0)
+    segments: Optional[int] = Field(description="Number of segments to draw, level of detail.")
+    theta_start: Optional[float] = Field(description="Start point position (relevant when this is an arc).")
+    theta_length: Optional[float] = Field(description="Values < PI*2 will result in an arc.")
+    rotation: Optional[Mat3[float]] = Field(description="Optional rotation of this item.")
```

Comment:
Should we replace this with a generic `ellipse` with `major_axis`, `minor_axis` and `normal` params and have builder methods that populate these fields by running some computations on the background?